File: resources/base/shared/logger.lua
---@alias colors 'white' | 'red' | 'green' | 'yellow' | 'blue' | 'light_blue' | 'purple' | 'white2' | 'orange' | 'grey' | number
local colors = {
    [0] = '^0', -- white
    [1] = '^1', -- red
    [2] = '^2', -- green
    [3] = '^3', -- yellow
    [4] = '^4', -- blue
    [5] = '^5', -- light_blue
    [6] = '^6', -- purple
    [7] = '^7', -- white2
    [8] = '^8', -- orange
    [9] = '^9', -- blue2
    white = '^0',
    red = '^1',
    green = '^2',
    yellow = '^3',
    blue = '^4',
    light_blue = '^5',
    purple = '^6',
    white2 = '^7',
    orange = '^8',
    blue2 = '^9',
}

Logger = {
    colors = colors
}

---@param text string
---@param color colors | nil
function Logger.colorize(text, color)
    return string.format('%s%s%s ', colors[color or 'white'], text, colors.white) or text
end

---@param prefix string
---@param color colors | nil
---@alias prefix { color: colors, prefix: string }
function Logger.prefix(prefix, color)
    return Logger.colorize(string.format('[%s]', prefix), color)
end

---@param message string
---@vararg prefix
function Logger.log(message, ...)
    local out = ''
    local args = table.pack(...)
    if #args > 0 then
        for i = 1, args.n do
            ---@type { color: colors, prefix: string }
            local prefix = args[i]
            out = out .. Logger.prefix(prefix.prefix, prefix.color)
        end
    end
    out = out .. message

    print(out)
    return out
end

---@param message string
---@param ... prefix
function Logger.error(message, ...)
    return Logger.log(message, { color = "red", prefix = "ERROR" }, ...)
end

---@param message string
---@param ... prefix
function Logger.info(message, ...)
    return Logger.log(message, { color = "blue", prefix = "INFO" }, ...)
end

---@param message string
---@param ... prefix
function Logger.success(message, ...)
    return Logger.log(message, { color = "green", prefix = "SUCCESS" }, ...)
end

---@param message string
---@param ... prefix
function Logger.verbose(message, ...)
    return Logger.log(message, { color = "purple", prefix = "VERBOSE" }, ...)
end

---@param message string
---@param ... prefix
function Logger.warn(message, ...)
    return Logger.log(message, { color = "yellow", prefix = "WARNING" }, ...)
end

Logger.info("INITIALIZED", { color = "yellow", prefix = "LOGGER" })

File: resources/base/src/shared/logger.ts
type Arg = { color: Color; prefix: string };
type Color = keyof typeof Logger.colors;

class Logger {
  static colors = {
    0: "^0",
    1: "^1",
    2: "^2",
    3: "^3",
    4: "^4",
    5: "^5",
    6: "^6",
    7: "^7",
    8: "^8",
    9: "^9",
    white: "^0",
    red: "^1",
    green: "^2",
    yellow: "^3",
    blue: "^4",
    light_blue: "^5",
    purple: "^6",
    white2: "^7",
    orange: "^8",
    blue2: "^9",
  };

  static colorize(text: string, color?: Color) {
    return `${this.colors[color || "white"]}${text}${this.colors.white}`;
  }

  static prefix(prefix: string, color?: Color) {
    return this.colorize(`[${prefix}] `, color || "white");
  }

  static log(message: string, ...args: Arg[]) {
    const log =
      args.reduce((prev, curr) => {
        return prev + this.prefix(curr.prefix, curr.color);
      }, "") + message;
    console.log(log);
    return log;
  }

  static error(message: string, ...args: Arg[]) {
    return this.log(message, { color: "red", prefix: "ERROR" }, ...args);
  }

  static info(message: string, ...args: Arg[]) {
    return this.log(message, { color: "green", prefix: "SUCCESS" }, ...args);
  }

  static verbose(message: string, ...args: Arg[]) {
    return this.log(message, { color: "purple", prefix: "VERBOSE" }, ...args);
  }

  static warn(message: string, ...args: Arg[]) {
    return this.log(message, { color: "yellow", prefix: "WARNING" }, ...args);
  }
}

Logger.info("INITIALIZED", { color: "yellow", prefix: "LOGGER" });
